File: scripts/vocabulary/translate-zh-to-vi.ts
/**
 * translate-zh-to-vi.ts
 *
 * Translates all Chinese (zh-CN) semantic content in dictionary files
 * to Vietnamese while preserving English words, phonetics, and metadata.
 *
 * Architecture:
 *   1. Rule-based POS label mapping (名→danh từ, etc.)
 *   2. For `cn` definition/translation fields, uses Gemini API if available,
 *      otherwise performs best-effort rule-based pass and flags for review.
 *   3. Preserves original Chinese in `cn_source` for traceability.
 *   4. Checkpoints per dictionary file for resumability.
 *
 * Usage:
 *   GEMINI_API_KEY=xxx node --strip-types scripts/vocabulary/translate-zh-to-vi.ts
 *   node --strip-types scripts/vocabulary/translate-zh-to-vi.ts           # rule-based only
 *   node --strip-types scripts/vocabulary/translate-zh-to-vi.ts --file CET4_T.json  # single file
 */

import fs from 'fs'
import path from 'path'

const PROJECT_ROOT = path.resolve(import.meta.dirname, '..', '..')
const DICT_DIR = path.resolve(PROJECT_ROOT, 'public/dicts/en/word')
const INVENTORY_PATH = path.resolve(import.meta.dirname, 'inventory.json')
const CHECKPOINT_DIR = path.resolve(import.meta.dirname, '.checkpoints')
const REPORT_PATH = path.resolve(import.meta.dirname, 'translation-report.json')

const GEMINI_API_KEY = process.env.GEMINI_API_KEY || ''
const GEMINI_MODEL = 'gemini-2.0-flash'
const GEMINI_ENDPOINT = `https://generativelanguage.googleapis.com/v1beta/models/${GEMINI_MODEL}:generateContent?key=${GEMINI_API_KEY}`
const BATCH_SIZE = 40 // words per Gemini API call
const API_DELAY_MS = 500

// --- Chinese detection ---
function containsChinese(text: string): boolean {
  return /[\u4e00-\u9fff\u3400-\u4dbf]/.test(text)
}

// --- POS mapping ---
const POS_MAP: Record<string, string> = {
  '名': 'danh từ', 'n.': 'd.', 'n': 'd.',
  '动': 'động từ', 'v.': 'đg.', 'v': 'đg.',
  'vt.': 'ngđ.', 'vi.': 'nđ.',
  '形': 'tính từ', 'adj.': 'tt.', 'a.': 'tt.',
  '副': 'trạng từ', 'adv.': 'trạng từ',
  '介': 'giới từ', 'prep.': 'gt.',
  '连': 'liên từ', 'conj.': 'lt.',
  '代': 'đại từ', 'pron.': 'đại từ',
  '数': 'số từ', 'num.': 'số từ',
  '冠': 'mạo từ', 'art.': 'mạo từ',
  '感': 'thán từ', 'interj.': 'thán từ',
  '助': 'trợ từ', '缩': 'viết tắt', 'abbr.': 'viết tắt',
  '叹': 'thán từ',
}

function translatePos(pos: string): string {
  if (!pos) return pos
  const trimmed = pos.trim()
  return POS_MAP[trimmed] || pos
}

// --- Gemini API translation ---
interface GeminiTranslation {
  index: number
  vi: string
}

async function translateBatchWithGemini(texts: { index: number; zh: string }[]): Promise<GeminiTranslation[]> {
  if (!GEMINI_API_KEY) return []

  const prompt = `You are a professional Chinese-Vietnamese dictionary translator for an English learning application.

Translate each numbered Chinese text below into natural, concise Vietnamese suitable for a dictionary entry.

Rules:
- Translate Chinese definitions into Vietnamese
- Do NOT translate English words that appear in the text
- Preserve semicolons (；) as separators between multiple meanings
- Use standard Vietnamese linguistic terminology
- Keep translations concise and dictionary-style
- For part-of-speech labels, use: danh từ, động từ, tính từ, trạng từ, giới từ, liên từ, đại từ
- For domain labels like [计] (computing), [医] (medicine), translate them: [tin học], [y học], etc.
- Return ONLY a JSON array of objects: [{"index": N, "vi": "translation"}, ...]

Texts to translate:
${texts.map(t => `[${t.index}] ${t.zh}`).join('\n')}

Return ONLY the JSON array, no markdown formatting.`

  try {
    const response = await fetch(GEMINI_ENDPOINT, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        contents: [{ parts: [{ text: prompt }] }],
        generationConfig: {
          temperature: 0.2,
          maxOutputTokens: 8192,
        },
      }),
    })

    if (!response.ok) {
      const errText = await response.text()
      throw new Error(`Gemini API ${response.status}: ${errText.slice(0, 200)}`)
    }

    const result = await response.json() as any
    const text = result.candidates?.[0]?.content?.parts?.[0]?.text || ''
    
    // Extract JSON from response (might be wrapped in markdown)
    const jsonMatch = text.match(/\[[\s\S]*\]/)
    if (!jsonMatch) {
      throw new Error(`Could not parse Gemini response: ${text.slice(0, 200)}`)
    }
    
    return JSON.parse(jsonMatch[0])
  } catch (err: any) {
    console.warn(`   ⚠️ Gemini API error: ${err.message}`)
    return []
  }
}

// --- Word translation ---
interface WordSubContent {
  c: string
  cn: string
  cn_source?: string
}

interface TransSense {
  pos: string
  cn: string
  cn_source?: string
  frequency?: number
}

interface SynoEntry {
  pos: string
  cn: string
  cn_source?: string
  ws: string[]
}

interface EtymologyEntry {
  t: string
  d: string
  t_source?: string
  d_source?: string
}

interface RelWords {
  root: string
  rels: {
    pos: string
    words: WordSubContent[]
  }[]
}

interface WordEntry {
  id?: string | number
  word: string
  phonetic0: string
  phonetic1: string
  trans: TransSense[]
  sentences: WordSubContent[]
  phrases: WordSubContent[]
  synos: SynoEntry[]
  relWords: RelWords
  etymology: EtymologyEntry[]
  [key: string]: any
}

interface TranslationStats {
  totalFields: number
  translatedByGemini: number
  translatedByRules: number
  preserved: number
  failed: number
}

function collectChineseTexts(word: WordEntry): { path: string; text: string }[] {
  const texts: { path: string; text: string }[] = []

  // trans[].cn
  word.trans?.forEach((t, i) => {
    if (t.cn && containsChinese(t.cn)) {
      texts.push({ path: `trans.${i}.cn`, text: t.cn })
    }
  })

  // sentences[].cn
  word.sentences?.forEach((s, i) => {
    if (s.cn && containsChinese(s.cn)) {
      texts.push({ path: `sentences.${i}.cn`, text: s.cn })
    }
  })

  // phrases[].cn
  word.phrases?.forEach((p, i) => {
    if (p.cn && containsChinese(p.cn)) {
      texts.push({ path: `phrases.${i}.cn`, text: p.cn })
    }
  })

  // synos[].cn
  word.synos?.forEach((s, i) => {
    if (s.cn && containsChinese(s.cn)) {
      texts.push({ path: `synos.${i}.cn`, text: s.cn })
    }
  })

  // relWords.rels[].words[].cn
  word.relWords?.rels?.forEach((r, i) => {
    r.words?.forEach((w, j) => {
      if (w.cn && containsChinese(w.cn)) {
        texts.push({ path: `relWords.rels.${i}.words.${j}.cn`, text: w.cn })
      }
    })
  })

  // etymology[].t and .d
  word.etymology?.forEach((e, i) => {
    if (e.t && containsChinese(e.t)) {
      texts.push({ path: `etymology.${i}.t`, text: e.t })
    }
    if (e.d && containsChinese(e.d)) {
      texts.push({ path: `etymology.${i}.d`, text: e.d })
    }
  })

  return texts
}

function applyTranslation(word: WordEntry, pathStr: string, translation: string): void {
  const parts = pathStr.split('.')
  let obj: any = word
  for (let i = 0; i < parts.length - 1; i++) {
    const key = parts[i]
    const idx = parseInt(key)
    obj = isNaN(idx) ? obj[key] : obj[idx]
  }
  const lastKey = parts[parts.length - 1]
  
  // Save source before overwriting
  if (lastKey === 'cn') {
    obj['cn_source'] = obj[lastKey]
  } else if (lastKey === 't' || lastKey === 'd') {
    obj[`${lastKey}_source`] = obj[lastKey]
  }
  
  obj[lastKey] = translation
}

async function translateDictionary(
  filePath: string,
  useGemini: boolean
): Promise<{ stats: TranslationStats; errors: string[] }> {
  const raw = fs.readFileSync(filePath, 'utf-8')
  const words: WordEntry[] = JSON.parse(raw)
  const stats: TranslationStats = {
    totalFields: 0,
    translatedByGemini: 0,
    translatedByRules: 0,
    preserved: 0,
    failed: 0,
  }
  const errors: string[] = []

  // Translate POS labels in trans[] (rule-based, always)
  for (const word of words) {
    if (word.trans) {
      for (const t of word.trans) {
        if (t.pos && containsChinese(t.pos)) {
          t.pos = translatePos(t.pos)
        }
      }
    }
    if (word.synos) {
      for (const s of word.synos) {
        if (s.pos && containsChinese(s.pos)) {
          s.pos = translatePos(s.pos)
        }
      }
    }
  }

  if (useGemini) {
    // Batch all Chinese texts across all words for Gemini translation
    const allTexts: { wordIdx: number; path: string; text: string; globalIdx: number }[] = []
    let globalIdx = 0
    
    for (let wi = 0; wi < words.length; wi++) {
      const chTexts = collectChineseTexts(words[wi])
      for (const ct of chTexts) {
        allTexts.push({ wordIdx: wi, path: ct.path, text: ct.text, globalIdx })
        globalIdx++
      }
    }
    
    stats.totalFields = allTexts.length
    console.log(`      ${allTexts.length} Chinese text fields to translate`)

    // Process in batches
    for (let batchStart = 0; batchStart < allTexts.length; batchStart += BATCH_SIZE) {
      const batch = allTexts.slice(batchStart, batchStart + BATCH_SIZE)
      const batchTexts = batch.map(t => ({ index: t.globalIdx, zh: t.text }))
      
      const translations = await translateBatchWithGemini(batchTexts)
      
      // Apply translations
      const translationMap = new Map(translations.map(t => [t.index, t.vi]))
      
      for (const item of batch) {
        const vi = translationMap.get(item.globalIdx)
        if (vi && vi.trim()) {
          applyTranslation(words[item.wordIdx], item.path, vi)
          stats.translatedByGemini++
        } else {
          // Keep original Chinese, mark as not translated
          stats.failed++
          if (errors.length < 20) {
            errors.push(`${words[item.wordIdx].word}:${item.path}`)
          }
        }
      }

      if (batchStart + BATCH_SIZE < allTexts.length) {
        await new Promise(r => setTimeout(r, API_DELAY_MS))
      }
      
      // Progress
      const done = Math.min(batchStart + BATCH_SIZE, allTexts.length)
      process.stdout.write(`\r      Translated: ${done}/${allTexts.length} fields`)
    }
    console.log() // newline after progress
  } else {
    // Rule-based only: preserve Chinese with cn_source marker
    for (const word of words) {
      const chTexts = collectChineseTexts(word)
      stats.totalFields += chTexts.length
      
      for (const ct of chTexts) {
        // For cn fields, save source and keep as-is (no translation without API)
        // The cn_source allows future translation passes
        const parts = ct.path.split('.')
        let obj: any = word
        for (let i = 0; i < parts.length - 1; i++) {
          const key = parts[i]
          const idx = parseInt(key)
          obj = isNaN(idx) ? obj[key] : obj[idx]
        }
        const lastKey = parts[parts.length - 1]
        
        if (lastKey === 'cn') {
          obj['cn_source'] = obj[lastKey]
        } else if (lastKey === 't' || lastKey === 'd') {
          obj[`${lastKey}_source`] = obj[lastKey]
        }
        stats.preserved++
      }
    }
  }

  // Write translated file
  fs.writeFileSync(filePath, JSON.stringify(words, null, 0), 'utf-8')
  
  return { stats, errors }
}

// --- Main ---
async function main() {
  const args = process.argv.slice(2)
  const singleFile = args.find((_, i) => args[i - 1] === '--file')
  const useGemini = !!GEMINI_API_KEY
  
  console.log('[TRANSLATE] TypeWords zh-CN → Vietnamese Translation')
  console.log(`   Translation engine: ${useGemini ? 'Gemini API (' + GEMINI_MODEL + ')' : 'Rule-based (POS labels only)'}`)
  console.log()

  if (!useGemini) {
    console.log('[TIP] For full translation, set GEMINI_API_KEY environment variable.')
    console.log('   Without it, only POS labels are translated; Chinese content is preserved with cn_source markers.')
    console.log()
  }

  // Load inventory
  if (!fs.existsSync(INVENTORY_PATH)) {
    console.error('[FAIL] Inventory not found. Run discover-source.ts first.')
    process.exit(1)
  }
  
  const inventory = JSON.parse(fs.readFileSync(INVENTORY_PATH, 'utf-8'))
  let filesToProcess: { id: string; url: string; name: string }[]

  if (singleFile) {
    filesToProcess = inventory.filter((e: any) => e.url === singleFile)
    if (filesToProcess.length === 0) {
      console.error(`[FAIL] File ${singleFile} not found in inventory`)
      process.exit(1)
    }
  } else {
    filesToProcess = inventory
  }

  fs.mkdirSync(CHECKPOINT_DIR, { recursive: true })

  const totalStats: TranslationStats = {
    totalFields: 0,
    translatedByGemini: 0,
    translatedByRules: 0,
    preserved: 0,
    failed: 0,
  }
  const allErrors: { dict: string; errors: string[] }[] = []
  let processed = 0

  for (const entry of filesToProcess) {
    const filePath = path.resolve(DICT_DIR, entry.url)
    const checkpointPath = path.resolve(CHECKPOINT_DIR, `${entry.url}.done`)
    
    if (fs.existsSync(checkpointPath) && !singleFile) {
      console.log(`⏭️  [${++processed}/${filesToProcess.length}] ${entry.name} — already translated`)
      continue
    }
    
    if (!fs.existsSync(filePath)) {
      console.log(`[FAIL] [${++processed}/${filesToProcess.length}] ${entry.name} — file not found`)
      continue
    }

    console.log(`[RECORD] [${++processed}/${filesToProcess.length}] ${entry.name}`)
    
    try {
      const { stats, errors } = await translateDictionary(filePath, useGemini)
      
      totalStats.totalFields += stats.totalFields
      totalStats.translatedByGemini += stats.translatedByGemini
      totalStats.translatedByRules += stats.translatedByRules
      totalStats.preserved += stats.preserved
      totalStats.failed += stats.failed
      
      if (errors.length > 0) {
        allErrors.push({ dict: entry.name, errors })
      }
      
      // Write checkpoint
      fs.writeFileSync(checkpointPath, new Date().toISOString(), 'utf-8')
      
      console.log(`   [OK] ${stats.totalFields} fields | Gemini: ${stats.translatedByGemini} | Rules: ${stats.translatedByRules} | Preserved: ${stats.preserved} | Failed: ${stats.failed}`)
    } catch (err: any) {
      console.error(`   [FAIL] Error: ${err.message}`)
      allErrors.push({ dict: entry.name, errors: [err.message] })
    }
  }

  // Summary
  console.log()
  console.log('[SUMMARY] Translation Summary:')
  console.log(`   Total Chinese fields:  ${totalStats.totalFields.toLocaleString()}`)
  console.log(`   Translated (Gemini):   ${totalStats.translatedByGemini.toLocaleString()}`)
  console.log(`   Translated (Rules):    ${totalStats.translatedByRules.toLocaleString()}`)
  console.log(`   Preserved (cn_source): ${totalStats.preserved.toLocaleString()}`)
  console.log(`   Failed:                ${totalStats.failed.toLocaleString()}`)

  // Write report
  fs.writeFileSync(REPORT_PATH, JSON.stringify({
    timestamp: new Date().toISOString(),
    engine: useGemini ? `gemini:${GEMINI_MODEL}` : 'rule-based',
    stats: totalStats,
    errors: allErrors,
  }, null, 2), 'utf-8')
  console.log(`   Report: ${REPORT_PATH}`)
}

main().catch(err => {
  console.error('[FAIL] Fatal error:', err)
  process.exit(1)
})

File: scripts/vocabulary/translate-zh-to-vi.py
#!/usr/bin/env python3
"""
translate-zh-to-vi.py

High-performance multi-threaded Chinese (zh-CN) -> Vietnamese batch translation
pipeline for TypeWords vocabulary dataset.

Features:
- Preserves original Chinese in `cn_source`, `t_source`, `d_source` for traceability.
- Preserves 100% of English headwords, phonetics, example sentences, phrases, and grammar.
- Translates Chinese POS tags to Vietnamese equivalents.
- Multi-threaded concurrent batch translation (4 workers, 70 lines per batch).
- Persistent disk cache (`.translation-cache.json`) shared across all dictionaries.
- Checkpoint-based resumption (`.checkpoints/<file>.done`).
- Validates JSON integrity before saving.

Usage:
  python3 scripts/vocabulary/translate-zh-to-vi.py --file nce-new-1.json
  python3 scripts/vocabulary/translate-zh-to-vi.py --priority
  python3 scripts/vocabulary/translate-zh-to-vi.py --all
"""

import os
import sys
import json
import time
import re
import html
import urllib.request
import urllib.parse
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import List, Dict, Set, Any, Tuple

PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..'))
DICT_DIR = os.path.join(PROJECT_ROOT, 'public', 'dicts', 'en', 'word')
CACHE_FILE = os.path.join(os.path.dirname(__file__), '.translation-cache.json')
CHECKPOINT_DIR = os.path.join(os.path.dirname(__file__), '.checkpoints')
REPORT_FILE = os.path.join(os.path.dirname(__file__), 'translation-report.json')

CHINESE_PATTERN = re.compile(r'[\u4e00-\u9fff\u3400-\u4dbf]')

DOMAIN_REPLACEMENTS = [
    ('[计]', '[tin học]'), ('[计算机]', '[tin học]'), ('[信息]', '[CNTT]'),
    ('[医]', '[y học]'), ('[生]', '[sinh học]'), ('[化]', '[hóa học]'),
    ('[物]', '[vật lý]'), ('[数]', '[toán học]'), ('[经]', '[kinh tế]'),
    ('[商]', '[thương mại]'), ('[法]', '[luật]'), ('[体]', '[thể thao]'),
    ('[地]', '[địa lý]'), ('[海]', '[hàng hải]'), ('[航]', '[hàng không]'),
    ('[机]', '[cơ khí]'), ('[电]', '[điện tử]'), ('[文]', '[văn học]'),
    ('[语]', '[ngôn ngữ]'), ('[植]', '[thực vật]'), ('[动]', '[động vật]'),
    ('[口]', '[khẩu ngữ]'), ('[书]', '[văn viết]'), ('[俚]', '[tiếng lóng]'),
    ('[俗]', '[thông tục]'),
    ('【名】', '(danh từ)'), ('【动】', '(động từ)'), ('【形】', '(tính từ)'),
    ('【副】', '(trạng từ)'), ('【介】', '(giới từ)'), ('【连】', '(liên từ)'),
]

POS_MAP = {
    '名': 'danh từ', 'n.': 'd.', 'n': 'd.',
    '动': 'động từ', 'v.': 'đg.', 'v': 'đg.',
    'vt.': 'ngđ.', 'vi.': 'nđ.',
    '形': 'tính từ', 'adj.': 'tt.', 'a.': 'tt.',
    '副': 'trạng từ', 'adv.': 'trạng từ',
    '介': 'giới từ', 'prep.': 'gt.',
    '连': 'liên từ', 'conj.': 'lt.',
    '代': 'đại từ', 'pron.': 'đại từ',
    '数': 'số từ', 'num.': 'số từ',
    '冠': 'mạo từ', 'art.': 'mạo từ',
    '感': 'thán từ', 'interj.': 'thán từ',
    '助': 'trợ từ', '缩': 'viết tắt', 'abbr.': 'viết tắt',
    '叹': 'thán từ',
}

PRIORITY_DICTS = [
    'CET4_T.json',
    'CET6_T.json',
    'IELTS_3_T.json',
    'TOEFL_3_T.json',
    'Level4luan_2_T.json',      # TEM-4 / 专四
    'Level8luan_2_T.json',      # TEM-8 / 专八
    'nce-new-1.json',
    'nce-new-2.json',
    'nce-new-3.json',
    'nce-new-4.json',
    'it-words.json',            # Programmer English / 程序员常用词
    'itVocabulary.json',        # Computer English / 计算机专用英语
    'KaoYan_3_T.json',          # 考研
    'GaoKao_3500.json',         # 高考 3500 词
    'CET-4-frequency.json',     # CET-4（翻译带频率）
]

def load_cache() -> Dict[str, str]:
    if os.path.exists(CACHE_FILE):
        try:
            with open(CACHE_FILE, 'r', encoding='utf-8') as f:
                return json.load(f)
        except Exception:
            pass
    return {}

def save_cache(cache: Dict[str, str]):
    tmp = CACHE_FILE + '.tmp'
    with open(tmp, 'w', encoding='utf-8') as f:
        json.dump(cache, f, ensure_ascii=False, indent=2)
    os.replace(tmp, CACHE_FILE)

def contains_chinese(text: Any) -> bool:
    if not isinstance(text, str):
        return False
    return bool(CHINESE_PATTERN.search(text))

def preprocess_chinese_text(text: str) -> str:
    res = text
    for ch, vi in DOMAIN_REPLACEMENTS:
        res = res.replace(ch, vi)
    return res

def translate_pos(pos: str) -> str:
    if not pos:
        return pos
    trimmed = pos.strip()
    return POS_MAP.get(trimmed, pos)

def call_google_translate_batch(texts: List[str], max_retries: int = 3) -> List[str]:
    if not texts:
        return []
    
    cleaned = [preprocess_chinese_text(t.replace('\r', '').replace('\n', ' ')).strip() for t in texts]
    joined = '\n'.join(cleaned)
    url = 'https://translate.google.com/m?sl=zh-CN&tl=vi&q=' + urllib.parse.quote(joined)
    
    headers = {
        'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 16_0 like Mac OS X) AppleWebKit/605.1.15'
    }
    
    for attempt in range(1, max_retries + 1):
        try:
            req = urllib.request.Request(url, headers=headers)
            with urllib.request.urlopen(req, timeout=14) as res:
                raw = res.read().decode('utf-8')
                m = re.search(r'<div class="result-container">([\s\S]*?)</div>', raw)
                if m:
                    unescaped = html.unescape(m.group(1)).strip()
                    lines = [l.strip() for l in unescaped.split('\n')]
                    if len(lines) == len(texts):
                        return lines
                    elif len(lines) > 0:
                        break
        except Exception:
            if attempt < max_retries:
                time.sleep(1.0 * attempt)
    
    # Fallback to single translations
    results = []
    for t in texts:
        results.append(translate_single(t))
    return results

def translate_single(text: str) -> str:
    if not contains_chinese(text):
        return text
    clean = preprocess_chinese_text(text.replace('\r', '').replace('\n', ' ')).strip()
    url = 'https://translate.google.com/m?sl=zh-CN&tl=vi&q=' + urllib.parse.quote(clean)
    headers = {'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 16_0 like Mac OS X)'}
    try:
        req = urllib.request.Request(url, headers=headers)
        with urllib.request.urlopen(req, timeout=8) as res:
            raw = res.read().decode('utf-8')
            m = re.search(r'<div class="result-container">([\s\S]*?)</div>', raw)
            if m:
                return html.unescape(m.group(1)).strip()
    except Exception:
        pass
    return text

def parallel_translate_unique(
    needed_strings: List[str],
    cache: Dict[str, str],
    batch_size: int = 70,
    max_workers: int = 4
) -> int:
    if not needed_strings:
        return 0
    
    batches = [needed_strings[i:i + batch_size] for i in range(0, len(needed_strings), batch_size)]
    completed = 0
    new_translated = 0
    total = len(needed_strings)
    
    print(f"      Translating {total} unique strings in {len(batches)} parallel batches ({max_workers} workers)...")
    
    with ThreadPoolExecutor(max_workers=max_workers) as executor:
        future_to_batch = {
            executor.submit(call_google_translate_batch, b): b for b in batches
        }
        
        last_save_time = time.time()
        for future in as_completed(future_to_batch):
            orig_batch = future_to_batch[future]
            try:
                results = future.result()
                for src, dst in zip(orig_batch, results):
                    if dst and dst != src:
                        cache[src] = dst
                        new_translated += 1
                    else:
                        cache[src] = src
            except Exception as e:
                for src in orig_batch:
                    cache[src] = src
                    
            completed += len(orig_batch)
            print(f"\r      Progress: {min(completed, total)}/{total} strings translated ({new_translated} new)", end='', flush=True)
            
            # Periodic cache save every 15s
            if time.time() - last_save_time > 15:
                save_cache(cache)
                last_save_time = time.time()

    print()
    save_cache(cache)
    return new_translated

def translate_dictionary_file(
    file_path: str,
    cache: Dict[str, str],
    batch_size: int = 70,
    max_workers: int = 4
) -> Tuple[int, int, int]:
    with open(file_path, 'r', encoding='utf-8') as f:
        words = json.load(f)
    
    unique_chinese: Set[str] = set()
    total_chinese_fields = 0
    
    for w in words:
        for t in w.get('trans', []):
            cn = t.get('cn', '')
            if contains_chinese(cn):
                unique_chinese.add(cn)
                total_chinese_fields += 1
        for s in w.get('sentences', []):
            cn = s.get('cn', '')
            if contains_chinese(cn):
                unique_chinese.add(cn)
                total_chinese_fields += 1
        for p in w.get('phrases', []):
            cn = p.get('cn', '')
            if contains_chinese(cn):
                unique_chinese.add(cn)
                total_chinese_fields += 1
        for s in w.get('synos', []):
            cn = s.get('cn', '')
            if contains_chinese(cn):
                unique_chinese.add(cn)
                total_chinese_fields += 1
        rel_words = w.get('relWords', {})
        if isinstance(rel_words, dict):
            for r in rel_words.get('rels', []):
                for sub_w in r.get('words', []):
                    cn = sub_w.get('cn', '')
                    if contains_chinese(cn):
                        unique_chinese.add(cn)
                        total_chinese_fields += 1
        for e in w.get('etymology', []):
            t_str = e.get('t', '')
            d_str = e.get('d', '')
            if contains_chinese(t_str):
                unique_chinese.add(t_str)
                total_chinese_fields += 1
            if contains_chinese(d_str):
                unique_chinese.add(d_str)
                total_chinese_fields += 1

    needed = [s for s in unique_chinese if s not in cache]
    cached_count = len(unique_chinese) - len(needed)
    
    translated_count = 0
    if needed:
        translated_count = parallel_translate_unique(needed, cache, batch_size=batch_size, max_workers=max_workers)
    else:
        print(f"      All {len(unique_chinese)} unique strings loaded from cache.")

    # Apply translations
    for w in words:
        for t in w.get('trans', []):
            if t.get('pos') and contains_chinese(t['pos']):
                t['pos'] = translate_pos(t['pos'])
            cn = t.get('cn', '')
            if contains_chinese(cn):
                t['cn_source'] = cn
                t['cn'] = cache.get(cn, cn)
                
        for s in w.get('sentences', []):
            cn = s.get('cn', '')
            if contains_chinese(cn):
                s['cn_source'] = cn
                s['cn'] = cache.get(cn, cn)
                
        for p in w.get('phrases', []):
            cn = p.get('cn', '')
            if contains_chinese(cn):
                p['cn_source'] = cn
                p['cn'] = cache.get(cn, cn)
                
        for s in w.get('synos', []):
            if s.get('pos') and contains_chinese(s['pos']):
                s['pos'] = translate_pos(s['pos'])
            cn = s.get('cn', '')
            if contains_chinese(cn):
                s['cn_source'] = cn
                s['cn'] = cache.get(cn, cn)
                
        rel_words = w.get('relWords', {})
        if isinstance(rel_words, dict):
            for r in rel_words.get('rels', []):
                if r.get('pos') and contains_chinese(r['pos']):
                    r['pos'] = translate_pos(r['pos'])
                for sub_w in r.get('words', []):
                    cn = sub_w.get('cn', '')
                    if contains_chinese(cn):
                        sub_w['cn_source'] = cn
                        sub_w['cn'] = cache.get(cn, cn)
                        
        for e in w.get('etymology', []):
            t_str = e.get('t', '')
            if contains_chinese(t_str):
                e['t_source'] = t_str
                e['t'] = cache.get(t_str, t_str)
            d_str = e.get('d', '')
            if contains_chinese(d_str):
                e['d_source'] = d_str
                e['d'] = cache.get(d_str, d_str)

    tmp_path = file_path + '.tmp'
    with open(tmp_path, 'w', encoding='utf-8') as f:
        json.dump(words, f, ensure_ascii=False)
    
    with open(tmp_path, 'r', encoding='utf-8') as f:
        reloaded = json.load(f)
        assert len(reloaded) == len(words), "Word count mismatch during write"
        
    os.replace(tmp_path, file_path)
    return (total_chinese_fields, translated_count, cached_count)

def main():
    args = sys.argv[1:]
    target_files = []
    
    if '--file' in args:
        idx = args.index('--file')
        if idx + 1 < len(args):
            target_files = [args[idx + 1]]
    elif '--priority' in args:
        target_files = PRIORITY_DICTS
    elif '--all' in args:
        target_files = sorted([f for f in os.listdir(DICT_DIR) if f.endswith('.json')])
    else:
        target_files = PRIORITY_DICTS

    os.makedirs(CHECKPOINT_DIR, exist_ok=True)
    cache = load_cache()
    print(f"[CACHE] Translation Cache loaded: {len(cache)} existing mappings")
    print(f"[TARGET] Target dictionaries: {len(target_files)}")
    print()

    total_fields = 0
    total_translated = 0
    success_count = 0

    for i, fname in enumerate(target_files, 1):
        fpath = os.path.join(DICT_DIR, fname)
        if not os.path.exists(fpath):
            print(f"⏭️  [{i}/{len(target_files)}] {fname} — file not found, skipping")
            continue
            
        chk = os.path.join(CHECKPOINT_DIR, f"{fname}.done")
        if os.path.exists(chk) and '--force' not in args:
            print(f"⏭️  [{i}/{len(target_files)}] {fname} — already translated (checkpoint exists)")
            continue

        print(f"[BOOK] [{i}/{len(target_files)}] Processing {fname}...")
        t0 = time.time()
        try:
            fields, trans, cached = translate_dictionary_file(fpath, cache)
            t1 = time.time()
            total_fields += fields
            total_translated += trans
            success_count += 1
            
            with open(chk, 'w', encoding='utf-8') as f:
                f.write(f"Done at {time.ctime()}, fields={fields}, trans={trans}, cached={cached}")
                
            print(f"   [OK] Done in {t1-t0:.1f}s: {fields} fields ({trans} new translated, {cached} cached)")
        except Exception as e:
            print(f"   [FAIL] Error processing {fname}: {e}")

    print()
    print("=" * 60)
    print(f"[SUCCESS] Completed {success_count}/{len(target_files)} dictionaries")
    print(f"[SUMMARY] Total Chinese fields processed: {total_fields}")
    print(f"[SAVE] Cache now contains {len(cache)} unique translation pairs")
    print("=" * 60)

if __name__ == '__main__':
    main()
